rare/models/game.py

```diff
@@ -0,0 +1,666 @@
+        if self.metadata.grant_date is None or force:
+            entitlements = self.core.lgd.entitlements
+            matching = filter(lambda ent: ent["namespace"] == self.game.namespace, entitlements)
+            entitlement = next(matching, None)
+            grant_date = datetime.fromisoformat(
+                entitlement["grantDate"].replace("Z", "+00:00")
+            ) if entitlement else None
+            if force:
+                print(grant_date)
+            self.metadata.grant_date = grant_date
+            self.__save_metadata()
+        return self.metadata.grant_date
+
+    @property
+    def can_launch(self) -> bool:
+        if self.is_installed:
+            if self.is_origin:
+                return True
+            if (not self.is_idle) or self.needs_verification:
+                return False
+            if self.is_foreign and not self.can_run_offline:
+                return False
+            return True
+        return False
+
+    def set_pixmap(self):
+        self.pixmap = self.image_manager.get_pixmap(self.app_name, self.is_installed)
+        if self.pixmap.isNull():
+            self.image_manager.download_image(self.game, self.set_pixmap, 0, False)
+        else:
+            self.signals.widget.update.emit()
+
+    def refresh_pixmap(self):
+        self.image_manager.download_image(self.game, self.set_pixmap, 0, True)
+
+    def install(self) -> bool:
+        if not self.is_idle:
+            return False
+        self.signals.game.install.emit(
+            InstallOptionsModel(app_name=self.app_name)
+        )
+        return True
+
+    def repair(self, repair_and_update):
+        self.signals.game.install.emit(
+            InstallOptionsModel(
+                app_name=self.app_name, repair_mode=True, repair_and_update=repair_and_update, update=repair_and_update
+            )
+        )
+
+    def uninstall(self) -> bool:
+        if not self.is_idle:
+            return False
+        self.signals.game.uninstall.emit(
+            UninstallOptionsModel(app_name=self.app_name)
+        )
+        return True
+
+    def launch(
+        self,
+        offline: bool = False,
+        skip_update_check: bool = False,
+        wine_bin: Optional[str] = None,
+        wine_pfx: Optional[str] = None,
+        ask_sync_saves: bool = False,
+    ) -> bool:
+        if not self.can_launch:
+            return False
+
+        cmd_line = get_rare_executable()
+        executable, args = cmd_line[0], cmd_line[1:]
+        args.extend(["start", self.app_name])
+        if offline:
+            args.append("--offline")
+        if skip_update_check:
+            args.append("--skip-update-check")
+        if wine_bin:
+            args.extend(["--wine-bin", wine_bin])
+        if wine_pfx:
+            args.extend(["--wine-prefix", wine_pfx])
+        if ask_sync_saves:
+            args.extend("--ask-sync-saves")
+
+        # kill me, if I don't change it before commit
```

Comment:
lol

In a more serious manner, you're absolutely awesome. It's quite an achievement to roll over with a 14k line diff PR during a whole year.

Reply:
The funnier part is that it was actually committed at some point and it remained for some time there, then it was fixed but the comment remained, and after that it was copied in that location with the comment still in place.

Reply:
Resolved by [a3d09ae](https://github.com/Dummerle/Rare/pull/259/commits/a3d09ae288b36de8f2810409bc3bb24ee4b72db7)